rasa/core/featurizers/single_state_featurizer.py

```diff
@@ -275,40 +309,36 @@ def encode_entities(
         Args:
             entity_data: The dict containing the text and entity labels and locations
             interpreter: The interpreter used to encode the state
+            bilou_tagging: indicates whether BILOU tagging should be used or not
 
         Returns:
             A dictionary of entity type to list of features.
         """
-        from rasa.nlu.test import determine_token_labels
-
         # TODO
         #  The entity states used to create the tag-idx-mapping contains the
         #  entities and the concatenated entity and roles/groups. We do not
         #  distinguish between entities and roles/groups right now.
-        # TODO
-        #  Should we support BILOU tagging?
-
-        if TEXT not in entity_data or len(self.entity_tag_id_mapping) < 2:
+        if (
+            not entity_data
+            or not self.entity_tag_specs
+            or self.entity_tag_specs[0].num_tags < 2
```

Comment:
Why not just have different instance variables for entities and roles/groups?

Reply:
because they don't always all appear and not always together